content/post/2023/14-solving-a-sudoku-with-symbiyosys-and-formal-verification.md

```diff
@@ -0,0 +1,305 @@
+---
+title: "Solving a Sudoku with SymbiYosys and Formal Verification"
+date: 2023-10-31T12:00:00+02:00
+description : Repurposing SymbiYosys to solve a Sudoku"
+tags: ["blog"]
+image: /static-2023/blinker-to-riscv.jpg
+slug: solving-sudoku-with-symbiyosys
+draft: true
+---
+
+_This guest post is by Theophile Loubiere._
+
+# Solving a Sudoku with SymbiYosys and Formal Verification
+
+Recently, I began using SymbiYosys to formally verify my designs. You can check out my first attempt on my blog [learn-fpga-easily](https://learn-fpga-easily.com/trying-formal-verification-with-symbiyosys-and-chisel/). Formal Verification helps ensure that certain properties of your design always remain true, such as:
+
+- **Bus arbitration**: "Only one master can receive the bus grant at any given time."
+- **Overflow and Underflow**: "The FIFO buffer will never overflow or underflow."
+- **State Machine Reachability**: "The state machine can never transition from state_1 to state_3."
+
+These kinds of properties would normally require an extensive functional testbench for coverage. However, with formal verification and SymbiYosys, they can be easily addressed with just a few assertions and assumptions.
+
+Given my newfound knowledge, it seemed completely natural that after embarrassingly failing to solve a Sudoku puzzle with my grandfather – a hit to my pride – I decided to hijack SymbiYosys to solve a Sudoku.
+
+Today, I am excited to share with you my overkill attempt to solve a simple Sudoku puzzle with SymbiYosys and Formal Verification.
+
+## Modeling the Sudoku
+
+In Verilog, we can represent a Sudoku grid straightforwardly: a two-dimensional register encompassing 9 rows and 9 columns, where each cell occupies 4 bits:
+
+``` verilog
+reg [3:0] sudoku_grid[8:0][8:0];
+```
+
+Given that Verilog lacks support for multi-dimensional module ports, it's necessary to flatten them. Additionally, to prevent our internal register from getting discarded during synthesis, we must link the two-dimensional register to the flattened output port.
```

Comment:
You can also use the `(* keep *)` attribute on a signal or register to prevent it from getting discarded and to make sure it occurs in the formal traces, even if it's not used anywhere.

If you do want to use an output port (which makes sense, it is the output after all) using a multi-dimensional array as an output port might work if you wrap the array in a struct as a packed array. (https://github.com/YosysHQ/yosys/blob/master/tests/svtypes/struct_array.sv has test cases of what yosys supports).

I think all 3 approaches, your current one, using (* keep *) or using a struct are all perfectly fine as they all showcase something different that you might end up also needing in other scenarios.

Reply:
Thanks for the information I choose the (* keep *) solution. It made me learn something